nm dns: Support appending static DNS before dynamic DNS

cathay4t opened the pull request:
Previously, nmstate does not allow storing DNS to auto interface with
`auto-dns: True`. This prevent the use case where user want to append
static DNS before dynamic DNS(for example, place 127.0.0.1 before DHCP
provided nameserver for DNS caching).

This patch will treat interface with `auto-dns: true` is valid to store
DNS, but not preferred.

Meanwhile, to save us from saving DNS to interface, this patch will try
to use NetworkManager Global DNS as much as possible unless:

 1. Desire state has IPv6 link-local nameserver, e.g. `fe80::deef:1%eth1`
 2. Desire state has static DNS and static IP interface.
 3. Desire state has static DNS with auto IP interface with
    `auto-dns: true` defined explicit.

To append DNS before dynamic DNS server, you need:

```yml
---
dns-resolver:
  config:
    search:
    - example.com
    - example.org
    server:
    - 8.8.8.8
    - 2001:4860:4860::8888
interfaces:
  - name: eth1
    type: ethernet
    state: up
    mtu: 1500
    ipv4:
      enabled: true
      dhcp: true
      auto-dns: true
    ipv6:
      enabled: true
      dhcp: true
      autoconf: true
      auto-dns: true
```

Integration test case included for use case on appending static DNS to
dynamic DNS nameservers.

liangwen12year reviewed (commented):
> Meanwhile, to save us from saving DNS to interface, this patch will try
> to use NetworkManager Global DNS as much as possible unless:

What if NM already had global DNS configured, will the DBus GlobalDnsConfiguration allow settings to be overwritten ?

cathay4t commented:
> > Meanwhile, to save us from saving DNS to interface, this patch will try
> > to use NetworkManager Global DNS as much as possible unless:
> 
> What if NM already had global DNS configured, will the DBus GlobalDnsConfiguration allow settings to be overwritten ?

Yes. And nmstate will always purge global DNS before making any changes:

```rust
    if merged_state.dns.is_changed()
        || !cur_dns_ifaces_still_valid_for_dns(&merged_state.interfaces)
    {
        purge_global_dns_config(&mut nm_api)?;
    }
```

cathay4t committed:
nm dns: Support appending static DNS before dynamic DNS

Previously, nmstate does not allow storing DNS to auto interface with
`auto-dns: True`. This prevent the use case where user want to append
static DNS before dynamic DNS(for example, place 127.0.0.1 before DHCP
provided nameserver for DNS caching).

This patch will treat interface with `auto-dns: true` is valid to store
DNS, but not preferred.

Meanwhile, to save us from saving DNS to interface, this patch will try
to use NetworkManager Global DNS as much as possible unless:

 1. Desire state has IPv6 link-local nameserver, e.g. `fe80::deef:1%eth1`
 2. Desire state has static DNS and static IP interface.
 3. Desire state has static DNS with auto IP interface with
    `auto-dns: true` defined explicit.

To append DNS before dynamic DNS server, you need:

```yml
---
dns-resolver:
  config:
    search:
    - example.com
    - example.org
    server:
    - 8.8.8.8
    - 2001:4860:4860::8888
interfaces:
  - name: eth1
    type: ethernet
    state: up
    mtu: 1500
    ipv4:
      enabled: true
      dhcp: true
      auto-dns: true
    ipv6:
      enabled: true
      dhcp: true
      autoconf: true
      auto-dns: true
```

Integration test case included for use case on appending static DNS to
dynamic DNS nameservers.

Signed-off-by: Gris Ge <fge@redhat.com>
